### PR TITLE
implement output layout auto configuration

### PR DESCRIPTION
**Pull request by acrisci**
#127

### CODE REVIEW

**Comment by acrisci** [comment]
To properly complete this implementation, I'm going to need to know whether or not an output is connected and get events when output connection status changes. The layout must be reconfigured when an output is disconnected. Right now, I'm hooking into the `output_remove` event from the backend and removing the output from the layout then (which triggers reconfiguration) but this can be handled automatically in a nicer way for users.

I think a disconnected output should rather stay within the layout but be invisible to any geometry calculations and queries.

I think we already discussed reasons for not wanting to put the backend within the high level api, so I think the events and connection queries should go on the wlr_output itself.

Connection status presently seems to be implementation specific, with the drm backend having its own enum for this. So I think we should add a function to the wlr_output interface called `wlr_output_is_connected()` that queries connection state.

As for who emits the connection status events, I think we have two options. One is to require the backend to emit the events on the output (essentially at the same time as the backend input_add/remove event) or for the backend to hook into its own input_add/remove event and emit the connection signal then. The latter is appealing because it takes some responsibility off of the backend.

**Comment by ascent12** [comment]
> Connection status presently seems to be implementation specific

It might be a good idea to move the DRM state-machine (that enum) into wlr_output. This was actually mentioned ages ago, around the time the DRM backend was first written, but nobody ever got around to it.

**Comment by SirCmpwn** [comment]
I prefer the approach where disconnected outputs are removed rather than bubbling the connector model up to the API surface.

**Comment by SirCmpwn** [comment]
To clarify, I don't think a disconnected output should stay in the layout. That would be weird.

**Comment by acrisci** [comment]
I got bored so I went ahead and did the implementation the way I described it to see what it looked like :)

As you see, it is a little bit weird. The nice thing is that you aren't required to re-add the output when it reconnects which is one less thing for users to keep track of, but I will defer to your judgement.

I think putting the DRM enum on the wlr_output is probably a good idea too.

**Comment by SirCmpwn** [comment]
The problem is that connectedness is a property of the connector, not the output. If you were to do this then there would have to be a wlr_connector which has ownership over the wlr_output. This would only be in use by the DRM backend and would add a fair amount of complexity.

The behavior you want is pretty weird, too. When a display is unplugged, its resolution and other properties become undefined until plugged in again, when it might have changed. People don't unplug their monitors and leave them there turned off, either - they unplug them and remove them from their desk, or unplug themselves from a projector, or undock their laptop and take it somewhere else.

:-1:

**Comment by ascent12** [comment]
>The nice thing is that you aren't required to re-add the output when it reconnects which is one less thing for users to keep track of

I don't think that's really a helpful feature. The way that we've designed it, the user should be able to able to handle displays being added and removed at any time. This somewhat specific special case needlessly breaks that "model".

Also, they can easily place that output in the right place, based on whatever their config is. We don't need to store that state for them.

File: types/wlr_output_layout.c [hunk 1]
```diff
@@ -47,16 +58,115 @@ void wlr_output_layout_destroy(struct wlr_output_layout *layout) {
...
+static struct wlr_box *wlr_output_layout_output_get_box(
+		struct wlr_output_layout_output *l_output) {
+		l_output->state->_box->x = l_output->x;
+		l_output->state->_box->y = l_output->y;
+		int width, height;
+		wlr_output_effective_resolution(l_output->output, &width, &height);
+		l_output->state->_box->width = width;
+		l_output->state->_box->height = height;
+		return l_output->state->_box;
+}
```
Comment (SirCmpwn) [inline types/wlr_output_layout.c]: Check indent

**Comment by SirCmpwn** [comment]
What's this new "main" output for? Also, we might want to talk about auto layouts with KDE.

**Comment by acrisci** [comment]
The "main" output exists only so the physics don't calculate at a different rate depending on how many monitors you have but it actually might be unnecessary.

**Comment by SirCmpwn** [comment]
Try to get rid of it, please.

**Comment by acrisci** [comment]
That all makes goos sense. You can consider this feature complete minus the last commit then.

**Comment by acrisci** [comment]
The "main" output here is necessary because otherwise, the animation speed will be proportional to the number of outputs which is not correct behavior. The problem is that each output has its own independent frame event so this gets called more times for more outputs.

This is actually going to be a problem for anyone who wants to do animations in the output frame callback (and where else would you do animations?). This solution is very simple and works for the example, but I believe that we will have to come up with some sort of framework to do animations. I think that's out of the scope of this PR, but definitely something that needs to be done.

**Comment by acrisci** [comment]
I think this is ready. This should fix #112 and #127. Please test hotplugging.

**Comment by ascent12** [comment]
> This is actually going to be a problem for anyone who wants to do animations in the output frame callback

I would actually consider that a bug in the application's code. Consistent frame events are not something we should promise, and I don't think other wayland compositors do that either. There are technologies like freesync which would also mess with that.

**Comment by SirCmpwn** [comment]
If you want to do animation in the frame callback you should get the current monotonic time and base it on that. Pull it out.

**Comment by acrisci** [comment]
I don't do many animation code so that wasn't immediately obvious to me like it should have been. Thanks.

**Comment by acrisci** [comment]
Ok that works much better.

**Comment by SirCmpwn** [comment]
Nice, thanks!

**Pull request merged**